odl/util/vectorization.py

```diff
@@ -0,0 +1,293 @@
+    ...     idcs = idcs[0] if len(idcs) > 1 else idcs
+    ...     out[idcs] = 1
+    ...     return out
+
+    Both versions work for arrays and meshgrids:
+
+    >>> x = np.linspace(-5, 13, 10, dtype=float).reshape((2, 5))
+    >>> x  # array representing 5 points in 2d
+    array([[ -5.,  -3.,  -1.,   1.,   3.],
+           [  5.,   7.,   9.,  11.,  13.]])
+    >>> np.array_equal(step(x), step_vec(x))
+    True
+
+    >>> x = y = np.linspace(-1, 2, 5)
+    >>> mg_sparse = np.meshgrid(x, y, indexing='ij', sparse=True)
+    >>> np.array_equal(step(mg_sparse), step_vec(mg_sparse))
+    True
+    >>> mg_dense = np.meshgrid(x, y, indexing='ij', sparse=False)
+    >>> np.array_equal(step(mg_dense), step_vec(mg_dense))
+    True
+
+    With output parameter:
+
+    >>> @vectorize(dtype=float)
+    ... def step(x):
+    ...     return 0 if x[0] <= 0 else 1
+    >>> x = np.linspace(-5, 13, 10, dtype=float).reshape((2, 5))
+    >>> out = np.empty(5, dtype=float)
+    >>> step(x, out)  # returns out
+    array([ 0.,  0.,  0.,  1.,  1.])
+    """
+    def vect_decorator(func):
+
+        def _vect_wrapper_array(x, out, **kwargs):
+            # Assume that x is an ndarray
+            if out is None:
+                out_shape = out_shape_from_array(x)
+                if dtype is None:
+                    out = [0] * out_shape[0]  # lazy vectorization
+                else:
+                    out = np.empty(out_shape, dtype=dtype)
+
+            for i, pt in enumerate(x.T):
+                out[i] = func(pt, **kwargs)
+            return np.asarray(out)
+
+        def _vect_wrapper_meshgrid(x, out, ndim, **kwargs):
+            if out is None:
+                out_shape = out_shape_from_meshgrid(x)
+
+                if dtype is None:
+
+                    # Small helper to initialize a nested list
+                    def _fill_nested(lst, shape):
+                        if len(shape) == 1:
+                            for _ in range(shape[0]):
+                                lst.append(float('nan'))
+                        else:
+                            for _ in range(shape[0]):
+                                lst.append([])
+                            for sublst in lst:
+                                _fill_nested(sublst, shape[1:])
+
+                    out = []
+                    _fill_nested(out, out_shape)
+
+                else:
+                    out = np.empty(out_shape, dtype=dtype)
+
+            order = meshgrid_input_order(x)
+            vecs = vecs_from_meshgrid(x, order=order)
+            if ndim == 1:
+                for i, pt in enumerate(vecs[0]):
+                    out[i] = func(pt, **kwargs)
+            elif dtype is None:
+                for i, pt in enumerate(product(*vecs)):
+                    # Worst multi-index implementation ever, but still good
+                    # enough for slow code
+                    index = np.unravel_index(i, out_shape)
```

Comment:
Shit, I wanted to avoid that question. However, we also default to `'C'` ordering when using arrays (this is the output side, so there's no well-defined order).